zookeeper-server/src/main/java/org/apache/zookeeper/server/PrepRequestProcessor.java

```diff
@@ -359,9 +355,6 @@ protected void pRequest2Txn(int type, long zxid, Request request, Record record,
         case OpCode.delete:
             zks.sessionTracker.checkSession(request.sessionId, request.getOwner());
             DeleteRequest deleteRequest = (DeleteRequest) record;
-            if (deserialize) {
```

Comment:
is this a behaviour change ? were is the deserialisation happening ?
are we doing it more times now ?

Reply:
No. Previously we need this boolean flag because, after deserializing `MultiOperationRecord`, the subrequest is an in-memory data structure - there are no more bytes in the request to be deserialized.

Now, we always deserialize in `pRequestHelper`, no serialization happens in `pRequest2Txn`. Besides, we always serialize at most once in `RequestRecord`:

```java
public class ByteBufferRequestRecord {
    @Override
    public <T extends Record> T readRecord(Supplier<T> constructor) throws IOException {
        if (record != null) {
            return (T) record;
        }
       // serialization
    }
}

public class SimpleRequestRecord {
    @Override
    public <T extends Record> T readRecord(Supplier<T> constructor) {
        return (T) record;
    }
}
```